solr/core/src/java/org/apache/solr/search/TermsQParserPlugin.java

```diff
@@ -88,7 +91,20 @@ Query makeFilter(String fname, BytesRef[] byteRefs) {
     docValuesTermsFilter {//on 4x this is FieldCacheTermsFilter but we use the 5x name any way
       @Override
       Query makeFilter(String fname, BytesRef[] byteRefs) {
-        return new DocValuesTermsQuery(fname, byteRefs);//constant scores
+        // TODO Further tune this heuristic number
+        return (byteRefs.length > 700) ? new TopLevelDocValuesTermsQuery(fname, byteRefs) : new DocValuesTermsQuery(fname, byteRefs);
+      }
+    },
+    docValuesTermsFilterTopLevel {
+      @Override
+      Query makeFilter(String fname, BytesRef[] byteRefs) {
+        return new TopLevelDocValuesTermsQuery(fname, byteRefs);
```

Comment:
I believe Joel argued strongly we should default both DV impls to cache=false?  You can do that here (and in the other method) easily via WrappedQuery wrapper and setting cache=false.  I mentioned another approach that messes with this.localParams but perhaps that might feel hacky compared to WrappedQuery.  Especially so since makeFilter doesn't have access to change the field this.localParams.

Reply:
I'd forgotten about this, done.

FWIW I buy Joel's argument that most real-world use-cases are going to require `cache=false`.  But I worry that users who want caching will be bitten by this.  I guess it depends how hands-on you expect users to be with their query tuning.  For hands-off users, this change is best. But for hands-on users, we've just created a "gotcha".

I'll make sure to add this to the ref-guide docs at the least.

Reply:
This is just about a default.  So-called "Hands-on users" (experts) will be able to be explicit with cache=true.

Reply:
Sure, setting`cache=true` is trivial.  It's _knowing_ that you need to be explicit for this one query in particular that's the tricky part.

But sure, I'm just hand-wringing; I've already made the change.  It's documented in the ref-guide, so that's as good as we're going to get.